### ISSUE TITLE
Cross-attention in ASD

### ISSUE THREAD

**Issue by HovChen** [Qinkaiyu/AOR-DR#3]
Hi, thanks for sharing the code!

While reading the paper and the implementation, I noticed that the current cross-attention fusion seems to compute attention across the batch dimension (i.e., sample-to-sample). In the paper, however, the cross-attention in ASD is described as operating within each sample, attending over the previous autoregressive prediction sequence ( $y_{1:j-1}$ ).

Thanks again for the nice work!

Best 🙂

**Comment by Qinkaiyu** [comment]
The fusion is applied within each sample, fusing the backbone features with the probability from the previous autoregressive prediction

**Comment by HovChen** [comment]
> The fusion is applied within each sample, fusing the backbone features with the probability from the previous autoregressive prediction

Hi @Qinkaiyu , thanks for the quick clarification!

I completely understand that the intended design is to perform fusion strictly within each sample.

However, I'd like to double-check the specific implementation in `SimpleFusion` (lines 75 in `main.py`). The code currently performs a transpose on the batch dimension:

```python
# Assuming features and context are both [Batch_Size, Embed_Dim]
attention_score = torch.matmul(features, context.transpose(0, 1)) 
# Calculation: [B, D] @ [D, B] -> Output is [B, B]
```

Mathematically, this results in a `[Batch_Size, Batch_Size]` matrix, where `attention_score[i, j]` represents the similarity between sample `i`'s feature and sample `j`'s context.

The subsequent softmax and multiplication:

```python
attention_weight = F.softmax(attention_score, dim=1)
attention_context = torch.matmul(attention_weight, context)
```

effectively computes a weighted sum of `context` from all samples in the batch for every single sample `i`. If strictly within-sample fusion is intended, this might inadvertently mix information across the batch.

Additionally, regarding the autoregressive input logic:

I noticed in the training loop (lines 370-378 in `main.py`), the code passes the output of the previous step (`fusion_feature`) as the input image feature for the next step:

```python
for t in range(4):
    if t == 0:
        # Step 0: Input is the original image feature
        loss, p, fusion_feature = feature_extractor(image, initial_p, vectors_targets[t], t)
    else:
        # Step t > 0: Input is 'fusion_feature' from the previous step
        loss, p, fusion_feature = feature_extractor(fusion_feature, p, vectors_targets[t], t)
```

This implementation recursively modifies the image feature at each step. However, Figure 1 and Equations (3) & (4) in the paper suggest that the fixed original global image feature ($z$) should be used as the condition at every step, combined with the updated prediction history.

Could you please confirm if this recursive feature update and the batch-wise attention are the intended behaviors?

Thanks again for your help!

**Comment by Qinkaiyu** [comment]
Hi, thanks a lot for the detailed and careful analysis — you’re absolutely right to point this out.
As for the autoregressive input logic, the paper’s formulation (Fig. 1 and Eqs. (3)(4)) is intentionally written in a clearer and more conceptual way, where the original global image feature 
𝑧 is used at every step. In the code, although the feature is recursively updated, we include residual connections, so the original image feature continuously propagates through the autoregressive steps. In practice, this leads to behavior that is functionally consistent with the paper’s description.

Regarding SimpleFusion, your understanding is correct. We agree this is a limitation of the current code, and in later jounal versions we switch to a strict per-sample cross-attention formulation to avoid this issue. We’ll update the public code accordingly soon.

**Comment by HovChen** [comment]
Hi, thanks for the explanation. I’ve updated the code to sample-wise fusion and kept the residual connections, but I still cannot reproduce the Table 1 results on the 4 datasets. The gap remains significant. To help me align with your original experiments, could you please share the checkpoints or the original script used for the paper results? This would be incredibly helpful for verifying the implementation. Thanks for your support!

**Comment by Qinkaiyu** [comment]
thanks for your message and for your careful attempts to reproduce the results.

Please kindly use the original code as provided in our GitHub repository. For the fusion module, you may follow either the affine transformation–based fusion (Lines 59–66) or the cross-attention–based fusion (Lines 68–78) as implemented in the released code.

I have provided the APTOS checkpoint (use cross-attn fusion) in the linked repository, which you can directly use for verification. In addition, I have recently updated the training logs for MESSIDOR and DEEPDR, which may be helpful for reference and alignment. https://drive.google.com/drive/folders/1AFiUMh3WB53XOMwYVm-2_6ooOcram77l?usp=sharing

I want to emphasize that the reported results are 100% reproducible with the released code and settings. One known limitation of the current version is that it requires a large number of training epochs to fully converge, which may partly explain the performance gap you observed. This issue has already been addressed in our journal extension, and I am currently organizing the updated code and pretrained weights. These will be released in the coming days, and I will inform you once they are available.

Thank you again for your interest and for your careful validation!

**Comment by HovChen** [comment]
Thanks a lot for your patience and support. I look forward to your updated code and the upcoming paper! 😊

**Comment by Qinkaiyu** [comment]
Hi @HovChen,

Thanks again for your patience. I’m glad to share that the updated journal-version code is now available. Here is the link
https://github.com/Qinkaiyu/DiffGeo-AOR

> Thanks a lot for your patience and support. I look forward to your updated code and the upcoming paper! 😊
